c2rust-analyze/src/known_fn.rs

```diff
@@ -602,28 +622,823 @@ pub const fn all_known_fns() -> &'static [KnownFn] {
+                // as it may be statically allocated and reused.  It is not meant to be modified.
+            ) -> *mut c_char: [READ | OFFSET_ADD];
+
+            #[cfg(target_os = "linux")]
+            fn setrlimit(
+                resource: __rlimit_resource_t,
+                rlim: *const rlimit: [READ | NON_NULL],
+            ) -> c_int;
+
+            #[cfg(target_os = "macos")]
+            fn setrlimit(
+                resource: c_int,
+                rlim: *const rlimit: [READ | NON_NULL],
+            ) -> c_int;
+
+            fn setsid() -> pid_t;
+
+            fn setsockopt(
+                socket: c_int,
+                level: c_int,
+                name: c_int,
+                value: *const c_void: [READ | NON_NULL],
+                option_len: socklen_t,
+            ) -> c_int;
+
+            fn setuid(
+                uid: uid_t,
+            ) -> c_int;
+
+            fn shutdown(
+                socket: c_int,
+                how: c_int,
+            ) -> c_int;
+
+            fn sigaction(
+                signum: c_int,
+                act: *const sigaction: [READ],
+                oldact: *mut sigaction: [WRITE],
+            ) -> c_int;
+
+            fn sigemptyset(
+                set: *mut sigset_t: [WRITE | NON_NULL],
+            ) -> c_int;
+
+            // fn signal(
+            //     signum: c_int,
+                   // TODO(kkysen) `libc::sighandler_t` is `size_t`, but may be a fn ptr.
+            //     handler: sighandler_t,
+            // ) -> sighandler_t;
+
+            // fn snprintf(
+            //     s: *mut c_char: [WRITE | OFFSET_ADD | NON_NULL],
+            //     n: size_t,
+            //     format: *const c_char: [READ | OFFSET_ADD | NON_NULL],
+                   // TODO(kkysen) varargs, printf-style
+            //     ...
+            // ) -> c_int;
+
+            fn socket(
+                domain: c_int,
+                ty: c_int,
+                protocol: c_int,
+            ) -> c_int;
+
+            #[cfg(target_os = "linux")]
+            fn splice(
+                fd_in: c_int,
+                off_in: *mut loff_t: [READ | WRITE],
+                fd_out: c_int,
+                off_out: *mut loff_t: [READ | WRITE],
+                len: size_t,
+                flags: c_uint,
+            ) -> ssize_t;
+
+            fn srand(
+                seed: c_uint,
+            ) -> ();
+
+            // TODO(kkysen) Not in `libc` crate.
+            // fn srandom;
+
+            fn stat(
+                path: *const c_char: [READ | OFFSET_ADD | NON_NULL],
+                buf: *mut stat: [WRITE | NON_NULL],
+            ) -> c_int;
+
+            fn strchr(
+                // `WRITE` because the return type is derived from `cs`'s provenance.
+                cs: *const c_char: [READ | WRITE | OFFSET_ADD | NON_NULL],
+                c: c_int,
+            ) -> *mut c_char: [READ | WRITE | OFFSET_ADD];
```

Comment:
At some point it might be nice to make functions like this pass-through certain permissions like `ptr::offset` does.  If the returned pointer is never written, there's no need to require `WRITE` on the argument.

Reply:
That would be very nice.  Basically being generic over mutability (and potentially other permissions, too), right?  We'd have to modify how we do `visit_local_call` with `LFnSig`s, though, so it'll be a ways away, and we'd have to add a way to annotate them as such.  It might make sense to have multiple declarations of the same function with different annotations, the way there are often `foo` and `foo_mut` `fn`s, and then resolve that at call sites (I was thinking of a similar thing for vararg `fn`s).  But that's definitely a question for later.  Want me to open an issue for it so we don't forget later on?

Reply:
Opened #1006.